pkg/controller/remotemachineset/remotemachineset_controller.go

```diff
@@ -213,85 +213,199 @@ func (r *ReconcileRemoteMachineSet) Reconcile(request reconcile.Request) (reconc
+	pool *hivev1.MachinePool,
+	cd *hivev1.ClusterDeployment,
+	remoteMachineSets *machineapi.MachineSetList,
+	logger log.FieldLogger,
+) ([]*machineapi.MachineSet, error) {
+	if pool.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	actuator, err := r.actuatorBuilder(cd, remoteMachineSets.Items, logger)
 	if err != nil {
-		cdLog.WithError(err).Error("unable to fetch remote machine sets")
-		return err
+		logger.WithError(err).Error("unable to create actuator")
+		return nil, err
 	}
-	cdLog.Infof("found %v remote machine sets", len(remoteMachineSets.Items))
 
-	generatedMachineSets := []*machineapi.MachineSet{}
-	machineSetsToDelete := []*machineapi.MachineSet{}
-	machineSetsToCreate := []*machineapi.MachineSet{}
-	machineSetsToUpdate := []*machineapi.MachineSet{}
+	// Generate expected MachineSets for Platform from InstallConfig
+	generatedMachineSets, err := actuator.GenerateMachineSets(cd, pool, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not generate machinesets")
+	}
 
-	if pool.DeletionTimestamp == nil {
-		actuator, err := r.actuatorBuilder(cd, remoteMachineSets.Items, cdLog)
-		if err != nil {
-			cdLog.WithError(err).Error("unable to create actuator")
-			return err
+	for i, ms := range generatedMachineSets {
+		if pool.Spec.Autoscaling != nil {
+			min, _ := getMinMaxReplicasForMachineSet(pool, generatedMachineSets, i)
+			ms.Spec.Replicas = &min
 		}
 
-		// Generate expected MachineSets for machine pool
-		generatedMachineSets, err = r.generateMachineSetsForMachinePool(cd, pool, actuator, cdLog)
-		if err != nil {
-			cdLog.WithError(err).Error("unable to generate machine sets for machine pool")
-			return err
+		if ms.Labels == nil {
+			ms.Labels = map[string]string{}
 		}
+		ms.Labels[machinePoolNameLabel] = pool.Spec.Name
 
-		cdLog.Infof("generated %v worker machine sets", len(generatedMachineSets))
+		// Apply hive MachinePool labels to MachineSet MachineSpec.
+		ms.Spec.Template.Spec.ObjectMeta.Labels = make(map[string]string, len(pool.Spec.Labels))
+		for key, value := range pool.Spec.Labels {
+			ms.Spec.Template.Spec.ObjectMeta.Labels[key] = value
+		}
 
-		// Find MachineSets that need updating/creating
-		for _, ms := range generatedMachineSets {
-			found := false
-			for _, rMS := range remoteMachineSets.Items {
-				if ms.Name == rMS.Name {
-					found = true
-					objectModified := false
-					objectMetaModified := false
-					resourcemerge.EnsureObjectMeta(&objectMetaModified, &rMS.ObjectMeta, ms.ObjectMeta)
-					msLog := cdLog.WithField("machineset", rMS.Name)
+		// Apply hive MachinePool taints to MachineSet MachineSpec.
+		ms.Spec.Template.Spec.Taints = pool.Spec.Taints
+	}
+
+	logger.Infof("generated %v worker machine sets", len(generatedMachineSets))
+
+	return generatedMachineSets, nil
+}
+
+func (r *ReconcileRemoteMachineSet) ensureEnoughReplicas(
+	pool *hivev1.MachinePool,
+	generatedMachineSets []*machineapi.MachineSet,
+	logger log.FieldLogger,
+) (*reconcile.Result, error) {
```

Comment:
Could use a comment on why this returns an optional result pointer, as well as an error, looks like it's to avoid a requeue if you just used errrors to stop the reconcile loop, as we know there's no point in re-reconciling here?

Reply:
Yes, it is so that we do not requeue in the case where the replicas count is too low. A requeue will never succeed until the machinepool is updated.

I will add a comment.